### PR TITLE
[encryption] fix moving files to a shared folder

### PR DESCRIPTION
**Pull request by schiessle**
Update share-keys if a file was moved to a shared folder.

Steps to test:
- user1 creates a folder called "folder1"
- user1 shares folder1 with user2
- user1 uploads a file "test.txt"
- the file will have one share-key for user1
- user1 moves test.txt fo folder1
- now test.txt should have two share-keys, one for user1 and one for user2 (doesn't work on master)
- repeat the same steps with webdav copy.

cc @PVince81 @th3fallen @DeepDiver1975

### CODE REVIEW

**Comment by PVince81** [comment]
Fix for https://github.com/owncloud/core/issues/15619

**Comment by PVince81** [comment]
- [x] move file into folder as owner with web UI
- [x] move file into folder as owner with WebDAV
- [x] move file into folder as owner with WebDAV and overwrite file
- [ ] copy file into folder as owner with WebDAV :no_entry_sign: FAIL: file still corrupted

**Comment by PVince81** [comment]
Oops, I actually tested it as recipient.
Testing again as owner now.

**Comment by PVince81** [comment]
@schiesbn I updated the test plan. Moving+copying as owner works.
But copy is broken as recipient. If you think it's not related to this PR (or you want to fix separately), please let me know and I'll raise a separate ticket.

**Comment by PVince81** [comment]
The copy issue also happens on master, will raise separately.

**Comment by PVince81** [comment]
Copy as recipient issue raised here: https://github.com/owncloud/core/issues/15840

:+1: this PR can be merged

**Comment by MorrisJobke** [comment]
If I move the file in the web UI I got `Cannot multikeyencrypt empty plain content` as yellow notification, but the file is moved to that folder (after a refresh/browse into folder).

**Comment by MorrisJobke** [comment]
This just happens with empty files and the welcome.txt. I couldn't reproduce this with a normal file.

**Comment by schiesbn** [comment]
@MorrisJobke the problem was that we tried to update the encryption key for non-encrypted files. Fixed with the last commit.

File: tests/lib/files/storage/wrapper/encryption.php [hunk 1]
```diff
@@ -80,28 +109,99 @@ public function setUp() {
...
+	 * @param boolean $shouldUpdate
+	 */
+	public function testRename($source, $target, $shouldUpdate) {
 		$this->keyStore
 			->expects($this->once())
-			->method('renameKeys');
-		$this->instance->mkdir('folder');
-		$this->instance->rename('folder', 'flodder');
+			->method('renameKeys')
+			->willReturn(true);
```
Comment (nickvergessen) [inline tests/lib/files/storage/wrapper/encryption.php]: Add a test where this returns false aswell?

File: tests/lib/files/storage/wrapper/encryption.php [hunk 2]
```diff
@@ -80,28 +109,99 @@ public function setUp() {
...
+	 *
+	 * @param string $source
+	 * @param string $target
+	 * @param boolean $shouldUpdate
+	 */
+	public function testCopy($source, $target, $shouldUpdate) {
+		$this->keyStore
+			->expects($this->once())
+			->method('copyKeys')
+			->willReturn(true);
```
Comment (nickvergessen) [inline tests/lib/files/storage/wrapper/encryption.php]: Add a test where this returns false aswell?

File: tests/lib/files/storage/wrapper/encryption.php [hunk 3]
```diff
@@ -80,28 +109,99 @@ public function setUp() {
...
+				->method('update');
+		}
+
+		$this->instance->mkdir($source);
+		$this->instance->mkdir(dirname($target));
+		$this->instance->copy($source, $target);
+	}
+
+	/**
+	 * data provider for testRename()
```
Comment (nickvergessen) [inline tests/lib/files/storage/wrapper/encryption.php]: wrong ;)

File: tests/lib/files/storage/wrapper/encryption.php [hunk 4]
```diff
@@ -80,28 +109,99 @@ public function setUp() {
...
+	 *
+	 * @return array
+	 */
+	public function dataTestCopy() {
```
Comment (nickvergessen) [inline tests/lib/files/storage/wrapper/encryption.php]: this one duplicates the other data provider... Just use the same one?

File: tests/lib/files/storage/wrapper/encryption.php [hunk 5]
```diff
@@ -13,26 +13,53 @@ class Encryption extends \Test\Files\Storage\Storage {
...
+	 */
+	private $encryptionModule;
+
+
+	/**
+	 * @var \OC\Encryption\Update | \PHPUnit_Framework_MockObject_MockObject
+	 */
+	private $update;
+
 	public function setUp() {
```
Comment (nickvergessen) [inline tests/lib/files/storage/wrapper/encryption.php]: `protected` :see_no_evil:

File: tests/lib/encryption/updatetest.php [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	/** @var \OC\Encryption\Manager | \PHPUnit_Framework_MockObject_MockObject */
+	private $encryptionManager;
+
+	/** @var \OCP\Encryption\IEncryptionModule | \PHPUnit_Framework_MockObject_MockObject */
+	private $encryptionModule;
+
+	/** @var \OC\Encryption\File | \PHPUnit_Framework_MockObject_MockObject */
+	private $fileHelper;
+
+	public function setUp() {
```
Comment (nickvergessen) [inline tests/lib/encryption/updatetest.php]: protected

File: lib/public/encryption/keys/istorage.php [hunk 7]
```diff
@@ -162,6 +163,7 @@ public function renameKeys($source, $target);
 	 *
 	 * @param string $source
 	 * @param string $target
+	 * @retrun boolean
```
Comment (nickvergessen) [inline lib/public/encryption/keys/istorage.php]: typo

File: lib/private/files/storage/wrapper/encryption.php [hunk 8]
```diff
@@ -66,15 +74,18 @@ class Encryption extends Wrapper {
...
 	 * @param string $uid user who perform the read/write operation (null for public access)
+	 * @param IStorage $keyStorage
+	 * @param Update $update
 	 */
 	public function __construct(
 			$parameters,
-			\OC\Encryption\Manager $encryptionManager = null,
-			\OC\Encryption\Util $util = null,
-			\OC\Log $logger = null,
+			Manager $encryptionManager = null,
```
Comment (nickvergessen) [inline lib/private/files/storage/wrapper/encryption.php]: Please type hint with the interface

File: lib/private/files/storage/wrapper/encryption.php [hunk 9]
```diff
@@ -66,15 +74,18 @@ class Encryption extends Wrapper {
...
+			Util $util = null,
+			Log $logger = null,
```
Comment (nickvergessen) [inline lib/private/files/storage/wrapper/encryption.php]: interface

**Comment by nickvergessen** [comment]
Added tests for the case @MorrisJobke mentioned and fixed some organisational stuff

:+1:

**Pull request merged**